File: packages/control-plane/src/diff.ts
// Manifest diff engine (spec 9.2 / 5.3), computed IN THE CONTROL PLANE.
//
// A run's result is a difference against its pre-run manifest (spec 5.3), and a
// fork difference view compares two manifests (spec 9.2). Both are functions of
// the manifests ALONE (decisions.md: "8.4 file browsing and 9.2 diffs are
// functions of the manifest alone") — no substrate, no wake, no chunk reads:
// entry metadata plus the per-file chunk-ref lists carry everything needed to
// decide added / modified / removed.
//
// The TS side only READS manifests (decisions.md); this module never writes one.

import type { DiffSummary, Manifest, ManifestEntry } from '@mari/shared';
import { loadManifest } from './manifest-store';

/** The metadata of one entry as it appears on one side of a diff. */
export interface DiffEntryFacts {
  kind: ManifestEntry['kind'];
  mode: number;
  size: number;
  symlink_target: string | null;
  /** Number of content chunks (0 for dirs/symlinks). */
  chunks: number;
}

/** An entry present on exactly one side. */
export interface DiffEntry extends DiffEntryFacts {
  path: string;
}

/** An entry present on both sides but not identical. */
export interface ModifiedEntry {
  path: string;
  from: DiffEntryFacts;
  to: DiffEntryFacts;
  /** `to.size - from.size`. */
  sizeDelta: number;
  /** The chunk-ref list changed (file content differs). */
  contentChanged: boolean;
  modeChanged: boolean;
  kindChanged: boolean;
  symlinkChanged: boolean;
}

/** The structured difference between two manifests (spec 9.2). */
export interface ManifestDiff {
  added: DiffEntry[];
  modified: ModifiedEntry[];
  removed: DiffEntry[];
  summary: DiffSummary;
}

function facts(e: ManifestEntry): DiffEntryFacts {
  return {
    kind: e.kind,
    mode: e.mode,
    size: e.size,
    symlink_target: e.symlink_target,
    chunks: e.chunks.length,
  };
}

/** True when two entries have identical content chunk refs (ids AND lengths). */
function sameChunks(a: ManifestEntry, b: ManifestEntry): boolean {
  if (a.chunks.length !== b.chunks.length) return false;
  for (let i = 0; i < a.chunks.length; i++) {
    const x = a.chunks[i]!;
    const y = b.chunks[i]!;
    if (x.chunk !== y.chunk || x.len !== y.len) return false;
  }
  return true;
}

function byPath(manifest: Manifest): Map<string, ManifestEntry> {
  const m = new Map<string, ManifestEntry>();
  for (const e of manifest.entries) m.set(e.path, e);
  return m;
}

/**
 * Diff `from` -> `to`. An entry is:
 *  - **added** when its path exists only in `to`,
 *  - **removed** when its path exists only in `from`,
 *  - **modified** when the path exists in both and any of kind/mode/size/
 *    symlink target/chunk refs differ.
 *
 * Results are path-sorted so the output is deterministic regardless of the
 * manifests' own ordering.
 */
export function diffManifests(from: Manifest, to: Manifest): ManifestDiff {
  const a = byPath(from);
  const b = byPath(to);

  const added: DiffEntry[] = [];
  const removed: DiffEntry[] = [];
  const modified: ModifiedEntry[] = [];

  for (const [path, toEntry] of b) {
    const fromEntry = a.get(path);
    if (!fromEntry) {
      added.push({ path, ...facts(toEntry) });
      continue;
    }
    const contentChanged = !sameChunks(fromEntry, toEntry) || fromEntry.size !== toEntry.size;
    const modeChanged = fromEntry.mode !== toEntry.mode;
    const kindChanged = fromEntry.kind !== toEntry.kind;
    const symlinkChanged = fromEntry.symlink_target !== toEntry.symlink_target;
    if (contentChanged || modeChanged || kindChanged || symlinkChanged) {
      modified.push({
        path,
        from: facts(fromEntry),
        to: facts(toEntry),
        sizeDelta: toEntry.size - fromEntry.size,
        contentChanged,
        modeChanged,
        kindChanged,
        symlinkChanged,
      });
    }
  }

  for (const [path, fromEntry] of a) {
    if (!b.has(path)) removed.push({ path, ...facts(fromEntry) });
  }

  const sortPath = <T extends { path: string }>(list: T[]): T[] =>
    list.sort((x, y) => (x.path < y.path ? -1 : x.path > y.path ? 1 : 0));

  return {
    added: sortPath(added),
    modified: sortPath(modified),
    removed: sortPath(removed),
    summary: { added: added.length, modified: modified.length, removed: removed.length },
  };
}

/** Total number of changed entries in a diff (the spec 8.2 "changed files"). */
export function changedCount(summary: DiffSummary): number {
  return summary.added + summary.modified + summary.removed;
}

/**
 * Load both manifests from the store and diff them (spec 9.2's engine, reading
 * through the manifest-store reader). Throws `ManifestNotFound` if either id is
 * absent.
 */
export async function diffManifestIds(
  store: R2Bucket,
  fromId: string,
  toId: string,
): Promise<ManifestDiff> {
  if (fromId === toId) {
    const only = await loadManifest(store, fromId);
    return diffManifests(only, only);
  }
  const [from, to] = await Promise.all([loadManifest(store, fromId), loadManifest(store, toId)]);
  return diffManifests(from, to);
}

File: crates/mari-core/src/diff.rs
//! Diff two manifests (spec 5.3, 9.2).
//!
//! A run's result, and a fork's difference view, are both functions of the
//! manifest alone — no filesystem access. Comparison is by path (both manifests
//! are path-sorted). A changed entry is classified so a mode-only change (chmod)
//! is distinguishable from a content change: the fork-diff must be able to show
//! "permissions changed" without claiming the bytes moved.

use mari_proto::{DiffSummary, ManifestEntry};

use mari_proto::Manifest;

/// One entry present in both manifests but changed.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ModifiedEntry {
    /// The shared path.
    pub path: String,
    /// The entry as it was in `a`.
    pub before: ManifestEntry,
    /// The entry as it is in `b`.
    pub after: ManifestEntry,
    /// The content changed: kind, size, chunk list, or symlink target differ.
    pub content_changed: bool,
    /// The mode (permission/type bits) changed.
    pub mode_changed: bool,
}

/// The full diff between two manifests, with per-entry detail.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ManifestDiff {
    /// Entries in `b` but not `a`.
    pub added: Vec<ManifestEntry>,
    /// Entries in `a` but not `b`.
    pub removed: Vec<ManifestEntry>,
    /// Entries in both, changed.
    pub modified: Vec<ModifiedEntry>,
}

impl ManifestDiff {
    /// The counts, as the wire [`DiffSummary`] (spec 5.3).
    pub fn summary(&self) -> DiffSummary {
        DiffSummary {
            added: self.added.len() as u32,
            modified: self.modified.len() as u32,
            removed: self.removed.len() as u32,
        }
    }

    /// True when the two manifests describe the same tree.
    pub fn is_empty(&self) -> bool {
        self.added.is_empty() && self.removed.is_empty() && self.modified.is_empty()
    }
}

/// Does the *content* (as opposed to just permissions) of two entries at the
/// same path differ? Symlink target and the ordered chunk list both count as
/// content.
fn content_differs(a: &ManifestEntry, b: &ManifestEntry) -> bool {
    a.kind != b.kind
        || a.size != b.size
        || a.symlink_target != b.symlink_target
        || a.chunks != b.chunks
}

/// Diff manifest `a` (before) against `b` (after). Both entry lists are assumed
/// path-sorted (snapshots always are); the walk is a linear merge by path.
pub fn diff(a: &Manifest, b: &Manifest) -> ManifestDiff {
    let mut out = ManifestDiff::default();
    let (av, bv) = (&a.entries, &b.entries);
    let (mut i, mut j) = (0usize, 0usize);
    while i < av.len() && j < bv.len() {
        let (ea, eb) = (&av[i], &bv[j]);
        match ea.path.cmp(&eb.path) {
            std::cmp::Ordering::Less => {
                out.removed.push(ea.clone());
                i += 1;
            }
            std::cmp::Ordering::Greater => {
                out.added.push(eb.clone());
                j += 1;
            }
            std::cmp::Ordering::Equal => {
                let content_changed = content_differs(ea, eb);
                let mode_changed = ea.mode != eb.mode;
                if content_changed || mode_changed {
                    out.modified.push(ModifiedEntry {
                        path: ea.path.clone(),
                        before: ea.clone(),
                        after: eb.clone(),
                        content_changed,
                        mode_changed,
                    });
                }
                i += 1;
                j += 1;
            }
        }
    }
    while i < av.len() {
        out.removed.push(av[i].clone());
        i += 1;
    }
    while j < bv.len() {
        out.added.push(bv[j].clone());
        j += 1;
    }
    out
}
